src/storage/indexeddb/storage.rs

```diff
@@ -0,0 +1,439 @@
+        let Some(cursor) = maybe_cursor else {
+            return Err(operation_not_found());
+        };
+
+        let found_op = serde_wasm_bindgen::from_value::<StoredOperation>(cursor.value()?)?;
+        if found_op.unsynced == 0 {
+            return Err(operation_synced());
+        }
+        if found_op.operation != op {
+            return Err(operation_not_found());
+        }
+        cursor.delete()?.await?;
+        Ok(())
+    }
+
+    async fn sync_complete(&mut self) -> Result<()> {
+        let ops = self.idb_txn()?.object_store(schema::OPERATIONS)?;
+        let ops_by_synced = ops.index(schema::OPERATIONS_BY_UNSYNCED)?;
+
+        // Update all operations to indicate they are sync'd. Using no query here returns only
+        // values with the `unsynced` property set.
+        let mut maybe_cursor = ops_by_synced.open_cursor(None, None)?.await?;
+        while let Some(cursor) = maybe_cursor {
+            let op = js2op(cursor.value()?)?;
+            cursor.update(&op2js(op, false)?)?.await?;
+            maybe_cursor = cursor.next(None)?.await?;
+        }
+
+        // Now delete all operations for which no task exists (usually deleted tasks).
+        let task_uuids: HashSet<Uuid, std::hash::RandomState> =
+            HashSet::from_iter(self.all_task_uuids().await?.drain(..));
+        let mut maybe_cursor = ops.open_cursor(None, None)?.await?;
+        while let Some(cursor) = maybe_cursor {
+            let stored_op = serde_wasm_bindgen::from_value::<StoredOperation>(cursor.value()?)?;
+            if let Some(uuid) = stored_op.uuid {
+                if !task_uuids.contains(&uuid) {
+                    cursor.delete()?.await?;
+                }
+            }
+            maybe_cursor = cursor.next(None)?.await?;
+        }
+
+        Ok(())
+    }
+
+    async fn get_working_set(&mut self) -> Result<Vec<Option<Uuid>>> {
+        let working_set = self.idb_txn()?.object_store(schema::WORKING_SET)?;
+        let mut res = vec![None];
+        let mut maybe_cursor = working_set
+            .open_cursor(None, Some(CursorDirection::Prev))?
+            .await?;
+        while let Some(cursor) = maybe_cursor {
+            let id = cursor.key()?.as_f64().ok_or_else(invalid)? as usize;
+            let uuid = js2uuid(cursor.value()?)?;
+            res.resize_with(res.len().max(id + 1), Default::default);
+            res[id] = Some(uuid);
+            maybe_cursor = cursor.next(None)?.await?;
+        }
+        Ok(res)
+    }
+
+    async fn add_to_working_set(&mut self, uuid: Uuid) -> Result<usize> {
+        let next_working_set_id = self.get_next_working_set_number().await?;
+        let working_set = self.idb_txn()?.object_store(schema::WORKING_SET)?;
+        working_set
+            .add(&uuid2js(uuid)?, Some(&next_working_set_id.into()))?
+            .await?;
+        Ok(next_working_set_id)
+    }
+
+    async fn set_working_set_item(&mut self, index: usize, uuid: Option<Uuid>) -> Result<()> {
```

Comment:
While the taskwarrior id is technically an index, I think this helps distinguish it from all the other "index" stuff going on here.

```suggestion
    async fn set_working_set_item(&mut self, id: usize, uuid: Option<Uuid>) -> Result<()> {
```